Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("https://drive.google.com/uc?export=download&id=1eNFlps9YaZZKsH4kgABg4iJUWTCBo98Y")

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

In [4]:
# проверим пропущенные знаечние
data.isna().sum()

age                        0
sex                        0
chest_pain_type            0
resting_blood_pressure     0
cholesterol                0
fasting_blood_sugar        0
rest_ecg                   0
max_heart_rate_achieved    0
exercise_induced_angina    0
st_depression              0
st_slope                   0
num_major_vessels          0
thalassemia                0
target                     0
dtype: int64

In [5]:
# посмотрим на типы данных 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
age                        303 non-null int64
sex                        303 non-null object
chest_pain_type            303 non-null object
resting_blood_pressure     303 non-null int64
cholesterol                303 non-null int64
fasting_blood_sugar        303 non-null object
rest_ecg                   303 non-null object
max_heart_rate_achieved    303 non-null int64
exercise_induced_angina    303 non-null object
st_depression              303 non-null float64
st_slope                   303 non-null object
num_major_vessels          303 non-null int64
thalassemia                303 non-null object
target                     303 non-null int64
dtypes: float64(1), int64(6), object(7)
memory usage: 33.2+ KB


<b>1. Сколько мужчин в датасете? Сколько женщен? (sex)</b>

In [6]:
data["sex"].value_counts()

male      207
female     96
Name: sex, dtype: int64

<b>Ответ:</b>

In [7]:
print("Количество мужчин: {}".format(data["sex"].value_counts()["male"]))
print("Количество женщин: {}".format(data["sex"].value_counts()["female"]))

Количество мужчин: 207
Количество женщин: 96


<b> 2. Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента)</b>

In [8]:
percent_male = round(((data[data["sex"] == "male"]["sex"].count())/(data["sex"].count()))*100, 2)

<b>Ответ:</b>

In [9]:
print("Процент мужчин в датасете: {} %".format(percent_male))

Процент мужчин в датасете: 68.32 %


<b> 3. Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?</b>

In [10]:
male_bad_heart = data[data["sex"] == "male"]["target"].sum()
female_bad_heart = data[data["sex"] == "female"]["target"].sum()

<b>Ответ:</b>

In [11]:
print("Количество мужчин с заболеванием сердца: {}".format(male_bad_heart))
print("Количество женщин с заболеванием сердца: {}".format(female_bad_heart))

Количество мужчин с заболеванием сердца: 93
Количество женщин с заболеванием сердца: 72


<b>4. Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?</b>

In [12]:
male_good_heart = data[(data["sex"] == "male")&(data["target"] == 0)]["sex"].count()
percent_male_good_heart = round(((male_good_heart)/(data["sex"].count()))*100, 2)

<b>Ответ:</b>

In [13]:
print("Доля мужчин, без болезни сердца, от общего числа пациентов: {} %".format(percent_male_good_heart))

Доля мужчин, без болезни сердца, от общего числа пациентов: 37.62 %


<b>5. Сколько лет самому молодому пациенту, страдающему болезнью сердца?</b>

In [14]:
youngest_bad_heart = data[data["target"] == 1]["age"].min()

<b>Ответ:</b>

In [15]:
print("Самому молодому пациенту, страдающему болезнью сердца - {} лет".format(youngest_bad_heart))

Самому молодому пациенту, страдающему болезнью сердца - 29 лет


<b>6. Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?</b>

In [16]:
oldest_no_bad_heart = data[data["target"] == 0]["age"].max()

<b>Ответ:</b>

In [17]:
print("Самому возрастному пациенту, у которого нет проблем с сердцем - {} лет".format(oldest_no_bad_heart))

Самому возрастному пациенту, у которого нет проблем с сердцем - 77 лет


<b>7. Сколько лет самой молодой женщине, которая страдает болезнью сердца?</b>

In [18]:
youngest_female_bad_heart = data[(data["sex"] == "female")&(data["target"] == 1)]["age"].min()

<b>Ответ:</b>

In [19]:
print("Самой молодой женщиен, которая страдает болезнью сердца - {} лет".format(youngest_female_bad_heart))

Самой молодой женщиен, которая страдает болезнью сердца - 34 лет


<b>8. Какой средний возраст женщин?</b>

In [20]:
avg_age_female = round(data[data["sex"] == "female"]["age"].mean(), 2)

<b>Ответ:</b>

In [21]:
print("Средний возраст женщин - {}".format(avg_age_female))

Средний возраст женщин - 55.68


<b>9. Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?</b>

In [22]:
avg_age_bad_heart = round(data[data["target"] == 1]["age"].mean(), 2)
std_age_bad_heart = round(np.sqrt(np.sum((data[data["target"] == 1]["age"] - avg_age_bad_heart)**2)/(data[data["target"] == 1].shape[0] - 1)), 2)

<b>Ответ:</b>

In [23]:
print("Среденее значение возраста тех, кто страдает болезнью сердца - {}".format(avg_age_bad_heart))
print("Ср. квадратичное отклонение возраста тех, кто страдает болезнью сердца - {}".format(std_age_bad_heart))

Среденее значение возраста тех, кто страдает болезнью сердца - 52.5
Ср. квадратичное отклонение возраста тех, кто страдает болезнью сердца - 9.55


<b>10. Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol)</b>

In [24]:
avg_cholesterol = round(data["cholesterol"].mean(), 2)
avg_cholesterol_good_heart = round(data[data["target"] == 0]["cholesterol"].mean(), 2)

<b>Ответ:</b>

In [25]:
print("Средний уровень холестерина: {}".format(avg_cholesterol))
print("Средний уровень холестерина, у тех, у кого нет проблем с сердцем: {}".format(avg_cholesterol_good_heart))

Средний уровень холестерина: 246.26
Средний уровень холестерина, у тех, у кого нет проблем с сердцем: 251.09


In [26]:
print("Не правда, у людей, не болеющие болезнью сердца, уровнь холестерина выше среднего")

Не правда, у людей, не болеющие болезнью сердца, уровнь холестерина выше среднего


<b> 11. Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)</b>

<b>Ответ:</b>

In [27]:
data.groupby(["rest_ecg"])[data.select_dtypes(include="number").drop(["target"], axis=1).columns].agg({"max": np.max,"mean": np.mean}).reset_index()

c:\users\vni\appdata\local\programs\python\python36\lib\site-packages\pandas\core\groupby\generic.py:1315: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


rest_ecg max                                     \
                                age resting_blood_pressure cholesterol   
0         ST-T wave abnormality  71                    180         354   
1  left ventricular hypertrophy  76                    180         327   
2                        normal  77                    200         564   

                                                                mean  \
  max_heart_rate_achieved st_depression num_major_vessels        age   
0                     194           5.6                 4  52.914474   
1                     140           4.4                 3  61.000000   
2                     202           6.2                 4  55.687075   

                                                                            \
  resting_blood_pressure cholesterol max_heart_rate_achieved st_depression   
0             129.065789  237.269737              151.960526      0.879605   
1             140.500000  261.750000              125.750000      2.725000   
2             134.027211  255.142857              147.904762      1.159184   

                     
  num_major_vessels  
0          0.638158  
1          1.000000  
2          0.816327

<b> 12. Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)</b>

In [30]:
dep_male_bad_heart = data[(data.sex == "male")&(data.target == 1)]["st_depression"].mean()
dep_female_good_heart = data[(data.sex == "female")&(data.target == 0)]["st_depression"].mean()

<b>Ответ:</b>

In [32]:
print("Ср. уровень депрессии при физ.нагрузке, среди мужчин, страдающих болезнью сердца: {}".format(dep_male_bad_heart))
print("Ср. уровень депресии при физ.нагрузке, среди женщин, не страдающих болезнью сердца: {}".format(dep_female_good_heart))

Ср. уровень депрессии при физ.нагрузке, среди мужчин, страдающих болезнью сердца: 0.6053763440860215
Ср. уровень депресии при физ.нагрузке, среди женщин, не страдающих болезнью сердца: 1.8416666666666668


In [33]:
print("Ср. уровень депрессии при физ.нагрузке, выше среди женщин не страдающих болезнью сердца")

Ср. уровень депрессии при физ.нагрузке, выше среди женщин не страдающих болезнью сердца


<b> 13. Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы</b>

<b>Ответ:</b>

In [35]:
data.groupby(["chest_pain_type"])["cholesterol"].agg({"max": np.max, "min": np.min})

c:\users\vni\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
  """Entry point for launching an IPython kernel.


max  min
chest_pain_type           
atypical angina   564  126
non-anginal pain  298  182
typical angina    409  131

In [37]:
data.groupby(["rest_ecg"])["cholesterol"].agg({"max": np.max, "min": np.min})

c:\users\vni\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
  """Entry point for launching an IPython kernel.


max  min
rest_ecg                              
ST-T wave abnormality         354  126
left ventricular hypertrophy  327  197
normal                        564  149

In [38]:
data.groupby(["thalassemia"])["cholesterol"].agg({"max": np.max, "min": np.min})

c:\users\vni\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
  """Entry point for launching an IPython kernel.


max  min
thalassemia                
fixed defect       417  141
normal             318  169
reversable defect  564  126

<b>14. Сколько значений может принимать каждый из категориальных признаков?</b>

In [46]:
cat_columns = data.select_dtypes(include="object").columns.tolist()
count_unique = {}
for cat in cat_columns:
    count_unique[cat] = data[cat].nunique()

<b>Ответ:</b>

In [52]:
print("Количество значений, которые может принимать категориальные признаки")
for key, value in count_unique.items():
    print(key, ":", value)

Количество значений, которые может принимать категориальные признаки
sex : 2
chest_pain_type : 3
fasting_blood_sugar : 2
rest_ecg : 3
exercise_induced_angina : 2
st_slope : 2
thalassemia : 3


<b>15. У какого категориального признака наблюдается самый сильный дизбаланс классов?</b>

In [56]:
d = dict(data["sex"].value_counts())

In [57]:
category = []
diff_between_max_min = []

for cat in cat_columns:
    count_values = dict(data[cat].value_counts())
    max_value = max(count_values.values())
    min_value = min(count_values.values())
    diff = max_value - min_value 
    category.append(cat)
    diff_between_max_min.append(diff)

cat_df = pd.DataFrame(data={"category": category, "difference": diff_between_max_min})

<b>Ответ:</b>

In [72]:
cat = cat_df[cat_df["difference"] == cat_df["difference"].max()]["category"].tolist()

print("Категориальный признак, с самым сильным дизбалансом: {}".format(cat[0]))

Категориальный признак, с самым сильным дизбалансом: fasting_blood_sugar


---

@Booster
Мамкин Data Scientist

tg @mommyscience